internal/credentials/xds/handshake_info.go

```diff
@@ -186,11 +236,69 @@ func (hi *HandshakeInfo) ServerSideTLSConfig(ctx context.Context) (*tls.Config,
 		if err != nil {
 			return nil, fmt.Errorf("xds: fetching trusted roots from CertificateProvider failed: %v", err)
 		}
-		cfg.ClientCAs = km.Roots
+		if km.SPIFFEBundleMap != nil {
+			// ClientAuth, if set above tls.RequireAnyClientCert, must be
+			// dropped to tls.RequireAnyClientCert so that custom verification
+			// to use SPIFFE Bundles is done.
+
+			if cfg.ClientAuth >= tls.VerifyClientCertIfGiven {
+				cfg.ClientAuth = tls.RequireAnyClientCert
+			}
+			cfg.VerifyPeerCertificate = hi.buildServerVerifyFunc(km)
+		} else {
+			cfg.ClientCAs = km.Roots
+		}
 	}
 	return cfg, nil
 }
 
+func (hi *HandshakeInfo) buildServerVerifyFunc(km *certprovider.KeyMaterial) func(rawCerts [][]byte, _ [][]*x509.Certificate) error {
+	return func(rawCerts [][]byte, _ [][]*x509.Certificate) error {
```

Comment:
There's a lot of duplication between this function and the client-side function. Can we consolidate this?